graylog2-server/src/main/java/org/graylog/security/DBGrantService.java

```diff
@@ -136,6 +141,27 @@ public GrantDTO create(GRN grantee, Capability capability, GRN target, String cr
         return create(GrantDTO.of(grantee, capability, target), creatorUsername);
     }
 
+    /**
+     * Ensure that a grant with the requested or a higher capability exists.
+     * @return the created, updated or existing grant
+     */
+    public GrantDTO ensure(GRN grantee, Capability capability, GRN target, String creatorUsername) {
+        final List<GrantDTO> existingGrants = getForTargetAndGrantee(target, grantee);
+        if (existingGrants.isEmpty()) {
+            return create(grantee, capability, target, creatorUsername);
+        }
+        // This should never happen
+        Preconditions.checkState(existingGrants.size() == 1);
+
+        final GrantDTO grantDTO = existingGrants.get(0);
+        // Only upgrade capabilities: VIEW < MANAGE < OWNER
+        if (capability.ordinal() > grantDTO.capability().ordinal()) {
```

Comment:
I am not sure if using the #ordinal method of the enum will bite us at one point in the future. The value will change if the enums are reordered at some point. (e.g. someone wants them to be sorted alphabetically and refactors the order) How about adding an explicit priority value to each enum instead?

Reply:
fair enough. :+1: